feat: [Data 2526] Adding documentation and license (#8)
hipagesgroup/hyperkops: README.md

# Hyperkops
© Hipages Group Pty Ltd 2019

This repository contains the components (collectively called Hyperkops) required to execute distributed Bayesian optimisation within
Kubernetes using the Python library [Hyperopt](https://github.com/hyperopt/hyperopt). 

The Hyperkops architecture is comprised of three main components:

* [Hyperkops Worker](https://hub.docker.com/r/hipages/hyperkops-worker) These run a hyperopt-worker, and execute each trial
* [Hyperkops Monitor](https://hub.docker.com/r/hipages/hyperkops-monitor): Identifies and updates hyperopt trials which should be logged as failed due to Pod failure or rotation
* [MongoDB](https://hub.docker.com/_/mongo): MongoDB Instance

An example helm chart can be found [here](https://github.com/hipagesgroup/chart-hyperopt).

# Hyperkops Architecture in Kubernetes

[Hyperopt](https://github.com/hyperopt/hyperopt) allows us to parallelise Bayesian optimisation jobs by distributing 
the experiments across multiple workers, with state stored and shared through a MongoDB instance. If this system is 

> operated on a conventional cluster, where the underlying infrastructure is expected to a longer lived than any optimisation

Reviewer: nah it should probably be
```suggestion
operated on a conventional cluster (where the underlying infrastructure is expected to be stable) is not expected to have hardware failures during any single job's lifetime
```

optimisation job. These can either be launched from your local machine, or from a Pod within Kubenernetes. 

# Example Workload
Provided [here](./examples/optimisation.py) is an example workload which matches that seen in the [Hyperopt Documentation](http://hyperopt.github.io/hyperopt/).
A prebuilt [Docker Container](https://hub.docker.com/r/hipages/hyperkops-example) is also provided, along with an example (Kubenetes Manifest)[./examples/kube-deploy-hyperkops-infrastructure.yaml],
 a Helm chart for this infrastructure is also [available](https://github.com/hipagesgroup/chart-hyperopt).

## Starting the infrastructure
  
### Helm
Helm users can find charts for infrastructure [here](https://github.com/hipagesgroup/chart-hyperopt). This can then be launched with the command:
Note defaults for the min and max number of works, and the autoscaling criteria may need to be changed to suit your use case, and cluster size.
### Kubernetes Manifest
There is also an example Kubernetes manifest to be found here, note that this launches jobs into a namespace of `datascience`. 

```kubectl apply -f kube-deploy-hyperkops-infrastructure.yam```

## Launching jobs from your local machine

> Typically connections external connections to pods within Kubernetes are handled by connecting their relevant service to an ingress. 

Reviewer: ```suggestion
Typically external connections to pods within Kubernetes are handled by connecting their relevant service to an ingress. 
```

Unfortunately, this doesn't work in Kubernetes because connections to the ingress pass through an Nginx instance which expects http connections. 
To get around these limitations its possible to connect to the MongoDB instance by port forwarding the relevant service 
within Kubernetes to your local instance. In our example this is done using:
```kubectl port-forward svc/hyperkops-mongo :27017:27017```

> Assuming you hve the correct privileges to port forward within your Kubernetes environment. 

Reviewer: ```suggestion
Assuming you have the correct privileges to port forward within your Kubernetes environment. 
```
